pintracker/maptracker/maptracker.go

```diff
@@ -201,53 +206,96 @@ func (mpt *MapPinTracker) isRemote(c api.Pin) bool {
 func (mpt *MapPinTracker) pin(c api.Pin) error {
 	logger.Debugf("issuing pin call for %s", c.Cid)
 	mpt.set(c.Cid, api.TrackerStatusPinning)
-	err := mpt.rpcClient.Call("",
+
+	var ctx context.Context
+	opc, ok := mpt.optracker.get(c.Cid)
+	if !ok {
+		logger.Debug("pin operation wasn't being tracked")
+		ctx = mpt.ctx
```

Comment:
I'm not quite seeing something here.

If the pin was queued, but the operation is cancelled (because an unpin request is received), this will still run when the pin comes out of the queue. Since the previous operation was cancelled, it was deleted and will show the warning that it wasn't tracked. Right?

Right now it would seem that it only cancels ongoing pins (by killing the context), but not queued pins. If an unpin requests comes and the pin is queued, the optracker should mark it as cancelled (phaseCancelled?), and this function should probably exit in that case.

This may force you to carry several operationCtx per Cid (one directly associated to each Track/Untrack) and have any new operation for that Cid affect the others when it runs (an unpin would cancel all pins). A new pin would probably not even enter the queue if there is a pin operationCtx for that Cid already

Probably, the operationCtx can only be removed from the tracker here, once it's no longer queued.

Please correct me if I missed something and things work differently.

Reply:
Thanks for catching this, I had forgotten to perform the 'filtering' of pin/unpins in the pin/unpinWorkers. I have pushed that change up. Please let me know if that doesn't solve the issue you are raising?

Reply:
I *think* it does solve it. Looking forward to tests.